tools/autograd/templates/variable_factories.h

```diff
@@ -22,61 +22,96 @@ using at::DimnameList;
-        type_(InitListTensorType::Scalar) {}
+  // 2. A Tensor represented in `std::vector<ListInitTensor>` form, with value
+  //    `vec()`, Tensor scalar type `scalar_type()`, and Tensor sizes `sizes()`.
+  struct ListInitTensor {
+#define TENSOR(T, S)                                                                     \
+    ListInitTensor(T scalar) :                                                           \
+        scalar_(scalar), vec_(),                                                         \
+        sizes_(),                                                                        \
+        scalar_type_(at::k##S),                                                          \
+        type_(ListInitTensorType::Scalar) {}                                             \
+    ListInitTensor(std::vector<T> values) :                                              \
+        scalar_(), vec_(),                                                               \
+        /* NOTE: We use `static_cast<int64_t>(...)` here because `std::vector<T>.size()`
+           returns a value of `std::vector<T>::size_type` type, and converting it to
+           `int64_t` type is a narrowing conversion. */                                  \
+        sizes_({static_cast<int64_t>(values.size())}),                                   \
+        scalar_type_(at::k##S),                                                          \
+        type_(ListInitTensorType::Vector) {                                              \
+      vec_.reserve(values.size());                                                       \
+      for (const auto& elem : values) {                                                  \
+        vec_.push_back(ListInitTensor(elem));                                            \
+      }                                                                                  \
+    }                                                                                    \
+    ListInitTensor(at::ArrayRef<T> values) :                                             \
+        scalar_(), vec_(),                                                               \
+        /* NOTE: We use `static_cast<int64_t>(...)` here because `at::ArrayRef<T>.size()`
+           returns a value of `size_t` type, and converting it to `int64_t` type is
+           a narrowing conversion. */                                                    \
+        sizes_({static_cast<int64_t>(values.size())}),                                   \
+        scalar_type_(at::k##S),                                                          \
+        type_(ListInitTensorType::Vector) {                                              \
+      vec_.reserve(values.size());                                                       \
+      for (const auto& elem : values) {                                                  \
+        vec_.push_back(ListInitTensor(elem));                                            \
```

Comment:
Discussed offline: there might be a performance regression here. For an ArrayRef with 1000 elements, before we would call `at::tensor`. Now the code calls `tensor.fill_` 1000 times. With 1us dispatch overhead this takes 6(!)ms, which is probably slower than what `at::tensor` can do (I imagine it does a memcpy).

Reply:
Oh, interesting. Previously, `torch::tensor({1, 2, 3, 4, 5, 6})` went through the ArrayRef ctor. Now, it goes through the ListInitTensor ctor. This means that there might be a performance regression for people who used to use `torch::tensor({1, 2, 3, 4, 5, 6})`